### DISCUSSION TITLE
Could Louvre be used to possibly port something like FluxBox to Wayland, or is it just a little too high-level?

### DISCUSSION THREAD

**Discussion by 8bitprodigy** [CuarzoSoftware/Louvre#29] [General]
Porting FluxBox to Wayland was part of why I started learning this project after all. My thought is to wrap Louvre API calls to something FluxBox can use to make it a bit more X11/WL-agnostic.

**Comment by ehopperdietzel** [comment]
I haven't had experience with FluxBox, so I'm not entirely sure about its features. Could you please inquire about the specific functionalities you're looking to implement? I'll then let you know if those are achievable with Louvre.

**Reply by 8bitprodigy** [reply]
Well, if you're familiar with NextStep's desktop, it works kinda like that. You can right click on the desktop to bring up their equivalent of a start menu(the root menu), as well as drag submenus around so they persist for quick access. 

I believe minimizing a window will either iconify or dock it. You can learn more about its features here: 
http://fluxbox.org/features/
And here:
https://en.m.wikipedia.org/wiki/Fluxbox

Essentially I want to try to port,  or maybe reimplement the features of fluxbox, if that turns out to be easier.

**Reply by ehopperdietzel** [reply]
Certainly, it seems entirely possible, and you have various methods to achieve this. Firstly, rendering is carried out within an `LOutput::paintGL()` event, where each screen (output) has its own rendering thread and `paintGL()` event. Within this event, you handle all the rendering for the upcoming frame to be displayed on the screen. You can use the `LPainter` functions, the `LScene` and `LView` system, or employ your own OpenGL ES 2.0 shaders or toolkits for rendering.

In Wayland, applications' windows are referred to as surfaces (`LSurface` in Louvre). Each surface has a buffer (OpenGL texture) which you can draw on screen as you wish. Surfaces also have roles, akin to window types, such as Toplevel for typical windows, Popups, Cursor (for cursor pixmaps that clients want to assign to the cursor), drag & drop icons, and subsurfaces, which are surfaces that can only be children of another surface and positioned relative to it.

If you wish to design your own decorations for toplevels, you can request that to the client (check the `LToplevelRole` class). However, not all clients support server-side decorations, for instance, GTK apps always use CSD. In such cases, you can request them to be maximized (while showing them in any size), effectively eliminating their decorations (something akin to what Sway does, I believe).

You also have access to the toplevels' window title string and app id string, which can be used to identify apps and display their text, for example, in a tab when minimized. However, for icons, you might need to adhere to freedesktop standards and possibly use systemd D-Bus services or a similar approach (which falls outside the scope of Wayland).

In summary, yes, you can achieve what you're looking for. I recommend reading the tutorial, as it provides insight into how everything is done and allows you to draw your own conclusions on how to implement it.

**Reply by 8bitprodigy** [reply]
Yeah, I'm going through the tutorial, currently. I was just wondering if this library could be used as like a replacement for the old X11 library, which fluxbox uses, and also makes creating a graphical environment fairly easy. Also,  looking at the fluxbox source code, it does appear to have its own toolkit, I assume for decorators and root menus: https://github.com/fluxbox/fluxbox

**Reply by ehopperdietzel** [reply]
Reviewing the code, it looks like Tk is being used, and I'm unsure if it supports an OpenGL ES 2.0 backend. In your position, I'd consider using LScene and LView variants. These classes efficiently render only what needs repainting in each frame. For example, they avoid painting content under opaque areas, or if the only change in an app is a button's texture, it only repaints that button. Configuring views is simple, just define their position, size, texture, opacity, etc.

I'd then recommend creating separate apps from the compositor using Qt or GTK for menus and UI elements. Then, establish communication between the compositor and these apps using either your own Wayland protocols, D-Bus or Unix domain sockets.

This approach would also enhance stability since, for any reason, if any of those components crash, the compositor wouldn't be affected.

Anyway, these are just suggestions. If you have doubts or need help, feel free to ask.

**Reply by 8bitprodigy** [reply]
Sorry for the delay, didn't see your reply.

I think the "Tk" in "FbTk" means "ToolKit", rather than referring to Tk, unless I'm missing an include somewhere. It looks like FluxBox defines its own UI toolkit that relies on the X11 libs. Think it's possible to replace those calls with Louvre calls?

Part of the reason for this effort was to avoid using Qt or GTK as they each have their own issues regarding management that I would like to avoid.

And yeah, I intend to separate them into different apps, which is ideal, as I'd like to be able to give users more customizability, andid liketobring in more BeOS/Haiku desktop features. Coincidentally and on that note, while I was going through the tutorial, I was thinking it would be cool to have a  tutorial showing how to make that top bar its own separate program that gets run when you run the Louvre-tutorial compositor.

Thanks for all your help so far!

**Reply by ehopperdietzel** [reply]
> I think the "Tk" in "FbTk" means "ToolKit", rather than referring to Tk, unless I'm missing an include somewhere. It looks like FluxBox defines its own UI toolkit that relies on the X11 libs. Think it's possible to replace those calls with Louvre calls?

Oops, you're right. I saw Tk and automatically assumed it was using that toolkit. Honestly, I suppose those calls could be replaced, but Louvre only provides functions for drawing textures and solid colors. So, for drawing more complex stuff, you would have to use your own shaders or toolkits. In that case, I would still recommend using LScene and LViews. Louvre offers various types of views, but you can create your own. It has several virtual methods that you can override, for example, [LView::paintRect()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_view.html#af72df78238280d5cccd8ddcef05873b8), which is called when the scene requests the view to be painted on the screen. You are free to paint it as you wish, so you can create your own toolkit that way. Furthermore, it has virtual methods to listen to input events.

> Part of the reason for this effort was to avoid using Qt or GTK as they each have their own issues regarding management that I would like to avoid.

I understand, I just mentioned them for the sake of naming something.

> And yeah, I intend to separate them into different apps, which is ideal, as I'd like to be able to give users more customizability, andid liketobring in more BeOS/Haiku desktop features. Coincidentally and on that note, while I was going through the tutorial, I was thinking it would be cool to have a tutorial showing how to make that top bar its own separate program that gets run when you run the Louvre-tutorial compositor.

Once I release Louvre 2.0 (which is still a long way off), I will continue developing a compositor I was working on called Crystals. It will have separate applications for the global menu, dock, desktop, etc. Therefore, I think it will be a good reference. For now, if you want to identify an app without using any additional protocol, you can assign the app ID of a window (toplevel) from the client. Then, you can retrieve that string from the compositor using [LToplevelRole::appId()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a1856f58a19ba32d243f03bb88d1de516) and listen for when the client changes the string by overriding the [LToplevelRole::appIdChanged()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a3fdd62206038c443119d34d69c67c8a1) method.

**Reply by 8bitprodigy** [reply]
> Louvre only provides functions for drawing textures and solid colors. So, for drawing more complex stuff, you would have to use your own shaders or toolkits.

So if I'm understanding correctly, there's a little bit of a gap in functionality between what Louvre provides and FluxBox needs, so I'd need to extend LScene and LViews in order to do this? Maybe I'll use Agar.

> For now, if you want to identify an app without using any additional protocol, you can assign the app ID of a window (toplevel) from the client. Then, you can retrieve that string from the compositor using [LToplevelRole::appId()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a1856f58a19ba32d243f03bb88d1de516) and listen for when the client changes the string by overriding the [LToplevelRole::appIdChanged()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a3fdd62206038c443119d34d69c67c8a1) method.

Thanks! So Louvre can be used for non-compositor programs to communicate with the compositor? I'm assuming this could be used to create a dock/taskbar/window menu.

Thanks again for all your help!

**Reply by ehopperdietzel** [reply]
> So if I'm understanding correctly, there's a little bit of a gap in functionality between what Louvre provides and FluxBox needs, so I'd need to extend LScene and LViews in order to do this? Maybe I'll use Agar.

Well, Louvre is designed for building compositors, and as its name implies, its purpose is to composite application windows (buffers) on screen which are accessible to you as OpenGL textures. Since it uses GLES 2.0, you could integrate Agar or any GLES 2.0 tools for more sophisticated rendering. However, if you aim for optimal compositor performance, alternative approaches could include creating an atlas of UI textures you want to use and displaying them as needed (rather than generating them or performing complex rendering in each frame with toolkits). Another approach would be to create Wayland clients (apps) to delegate such tasks (for docks, menus, etc.), and then simply display their generated textures.

> Thanks! So Louvre can be used for non-compositor programs to communicate with the compositor? I'm assuming this could be used to create a dock/taskbar/window menu.

Essentially, Wayland facilitates communication between clients (apps) and the compositor (display server) through Unix domain sockets. Everything else, such as rendering and accessing system input, must be implemented independently by the compositor itself. You can conceptualize a Wayland compositor as a web server with a REST API, where clients communicate with it, with the distinction that the server can also send messages to clients at any time. Now, there are [numerous protocols](https://wayland.app/protocols/) like the Wayland protocol, XDG Shell, Viewport, etc., and you even have the flexibility to create your own. Each protocol consists of a set of interfaces, and each interface has a set of requests and events. As the name implies, each protocol defines rules on how the compositor and clients should behave. Louvre implements some of these protocols, particularly the crucial ones, and provides a simplified C++ API to the dev. 

For example, when a client makes [this request](https://wayland.app/protocols/xdg-shell#xdg_toplevel:request:set_app_id) from the XDG Shell protocol and xdg_toplevel interface, Louvre first handles it in a low level layer [here](https://github.com/CuarzoSoftware/Louvre/blob/main/src/lib/protocols/XdgShell/private/RXdgToplevelPrivate.cpp), and then notifies to the public API invoking [LToplevelRole::appIdChanged()](https://cuarzosoftware.github.io/Louvre/class_louvre_1_1_l_toplevel_role.html#a3fdd62206038c443119d34d69c67c8a1).

You can also add your own protocols in Louvre (for example, to send custom events to clients or listen to custom requests from clients), you should of course implement such protocols in the client side as well. I recommend reading the [Wayland Book](https://wayland-book.com/) to gain a better understanding of how Wayland operates and how to create and implement your own protocols.

I should add a chapter to the tutorial to demonstrate how to create and implement a custom protocol both in Louvre and on the client side.

**Reply by 8bitprodigy** [reply]
Thanks again for your time helping me. I think if I do try to progress with this project, I'll make those parts that are in Fluxbox part of this one, too, at least at first, with the goal being to separate them later. I'll also have to look into what all xlib does that fluxbox uses.

**Reply by ehopperdietzel** [reply]
Fantastic! Best of luck 😃 Feel free to showcase your project in the [Gallery](https://github.com/CuarzoSoftware/Louvre/discussions/categories/gallery) if you choose to proceed, and don't hesitate to initiate new discussions for any specific questions you may encounter along the way.

**Reply by 8bitprodigy** [reply]
Will do. Also, there is currently waybox by wizbright which replicates OpenBox, which is similar, but using wlroots.
That said, it does link to some projects that can replicate some of those features in fluxbox, such as the panel and what-not.
